src/ripple/app/main/GRPCServer.cpp

```diff
@@ -0,0 +1,270 @@
+    {
+        auto usage = getUsage();
+        if(usage.disconnect())
+        {
+            grpc::Status status{grpc::StatusCode::RESOURCE_EXHAUSTED,
+                "usage balance exceeds threshhold"};
+            responder_.FinishWithError(status, this);
+        }
+        else
+        {
+
+            auto loadType = getLoadType();
+            usage.charge(loadType);
+            auto role = getRole();
+
+            RPC::ContextGeneric<Request> context {
+                app_.journal("gRPCServer"), request_, app_, loadType,
+                    app_.getOPs(), app_.getLedgerMaster(), usage, role, coro,
+                    InfoSub::pointer()};
+
+            //Make sure we can currently handle the rpc
+            error_code_i condition_met_res = 
+                RPC::conditionMet(required_condition_, context);
+
+            if(condition_met_res != rpcSUCCESS)
+            {
+                RPC::ErrorInfo error_info =
+                    RPC::get_error_info(condition_met_res);
+                grpc::Status status{grpc::StatusCode::INTERNAL,
+                    error_info.message.c_str()};
+                responder_.FinishWithError(status,this);
+            }
+            else
+            {
+                std::pair<Response,grpc::Status> result = handler_(context);
+                responder_.Finish(result.first, result.second, this);
+            }
+        }
+    }
+    catch(std::exception const & ex)
+    {
+        grpc::Status status{grpc::StatusCode::INTERNAL,ex.what()};
+        responder_.FinishWithError(status,this);
+    }
+}
+
+
+
+
+void GRPCServerImpl::handleRpcs() {
+    void* tag;  // uniquely identifies a request.
+    bool ok;
+    // Block waiting to read the next event from the completion queue. The
+    // event is uniquely identified by its tag, which in this case is the
+    // memory address of a CallData instance.
+    // The return value of Next should always be checked. This return value
+    // tells us whether there is any kind of event or cq_ is shutting down.
+    while (cq_->Next(&tag,&ok)) {
+
+        //if ok is false, event was terminated as part of a shutdown sequence
+        //need to abort any further processing
+        if(!ok)
+        {
+            //abort first, then erase. Otherwise, erase can delete object
+            static_cast<Processor*>(tag)->abort();
+            requests_.erase(static_cast<Processor*>(tag)->get_iter());
+        }
+        else
+        {
+            auto ptr = static_cast<Processor*>(tag);
+            if(!ptr->isFinished())
+            {
+                //ptr is now processing a request, so create a new CallData
+                //object to handle additional requests
+                auto cloned = ptr->clone();
+                requests_.push_front(cloned);
+                //set iterator as data member for later lookup
+                cloned->set_iter(requests_.begin());
+                //process the request
+                ptr->process();
```

Comment:
We are using shared pointers here, do we have to? It looks like the lifetime of the processors are controlled by the `requests_` collection (am I right in that?). If so, why not store the raw pointers in `requests_` and then delete them when they are removed?

If we do this, I'd make some other changes as well (I'll add additional comments where the other changes could be made as well):
1) `clone` can return a raw pointer
2) `CallData` no longer needs to inherit from `enable_shared_from_this`

Reply:
The shared pointer is because the object can be accessed from two different threads: 
1. the thread that's running this function (`handleRpcs()`) 
2. the thread that executes the coroutine posted to the JobQueue (which calls the actual handler).

There could be concurrent access to this object when the server shuts down. When the server shuts down, all of the `CallData` objects are erased from the `requests_` collection, but the coroutine may not have executed yet. We're going to have undefined behavior if the coroutine executes after the `CallData` object is deleted. So, I'm using shared pointers to ensure the object is not deleted until after the coroutine executes (note that in the event of server shutdown, the coroutine will return immediately since the `aborted_` member will be set to true)

Reply:
Ah, I thought the `handleRpcs` wouldn't exit until after the job queue finished (I didn't dig into `cq_->next` yet). I agree, we need the shared_pointers. Not a big deal. I still think we can make get rid of `enable_shared_from_this` by passing it in as a parameter. I'm fine if you want to keep the return type of `clone` as a shared_ptr tho. That makes sense given how it's used.

Reply:
Just to be clear, you are proposing making `process()` static, and passing the `shared_ptr` as a parameter to that `process()`, right? I agree that would work. Is there a reason that would be preferred over `enable_shared_from_this`?

Reply:
@cjcobb23 - yes, that's what I was proposing. I was thinking it's easier to following lifetime issues if member functions don't create shared pointers themselves that get pushed onto the job queue. However, we pass parameters using "shared_from_this" stuff in member functions in asio code all the time. It's fine as-is, no change needed.

Reply:
kept this as is